zict/tests/test_lru.py

```diff
@@ -88,6 +91,25 @@ def cb(k, v):
     assert count[0] == len(L)
 
 
+def test_cb_exception_keep_on_lru():
+    def cb(k, v):
+        raise Exception
+
+    L = list()
+    d = dict()
+    lru = LRU(2, d, on_evict=[lambda k, v: L.append((k, v)), cb])
+
+    lru["x"] = 1
+    lru["y"] = 2
+
+    with pytest.raises(Exception):
+        lru["z"] = 3
+
+    assert L == [("x", 1)]  # tried to evict and raised exception
+    assert lru.total_weight == 3
+    assert lru.weights == {"x": 1, "y": 1, "z": 1}
+
```

Comment:
+ Please add check (black box preferably) that y and z are safely stored in self.d
+ Please add check (black box preferably) that the heap hasn't been corrupted

Reply:
> Please add check (black box preferably) that y and z are safely stored in self.d (suggestion: you can do it in the test above by setting weight=lambda k, v: v)

I added a separate test that uses `wight` since in LRU `n` will mean two different things depending on whether we provide weights or not. 
```
n: int
        Number of elements to keep, or total weight if weight= is used
```

I'm not sure if I capture what you meant by "check that y and z are stored safely. " let me know if I misunderstood you. 

> Please add check (black box preferably) that the heap hasn't been corrupted

I wasn't quite sure what you meant by "corrupted" here, I'm checking now that the keys are in the heap with the right priority. 

> Please test use case where individual weight >= n

I added a case to check weight >= n

All changes are in https://github.com/dask/zict/pull/48/commits/cab104171cbe08099231268181689899e504af84